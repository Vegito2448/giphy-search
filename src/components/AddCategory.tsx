import { ChangeEvent, FormEventHandler, useState } from "react";

interface AddCategoryProps {
  setCategories: React.Dispatch<React.SetStateAction<Array<string>>>;
}

export const AddCategory = ({ setCategories }: AddCategoryProps) => {

  const [inputValue, setInputValue] = useState('');


  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
  };

  const handleSubmit: FormEventHandler<HTMLFormElement> = (e) => {
    e.preventDefault();
    if (inputValue.trim().length > 2) {
      setCategories(prev => ([inputValue, ...prev]));
      setInputValue('');
    }
  };

  return (
    <form onSubmit={handleSubmit} >
      <input
        type="search"
        onChange={handleInputChange}
        value={inputValue}
        style={{ marginRight: '10px' }}

      />
      <button type="submit">Search</button>
    </form>
  );
};
