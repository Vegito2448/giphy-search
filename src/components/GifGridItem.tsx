
import 'animate.css';
interface GifGridItemProps {
  id: string;
  title: string;
  url: string;
}

export const GifGridItem = ({ url, title }: GifGridItemProps) => {
  return (
    <div
      className="card animate__animated animate__fadeIn animate__delay-2s"
    >
      <img src={url} alt={title} />
      <p>{title}</p>
    </div>
  );
};
