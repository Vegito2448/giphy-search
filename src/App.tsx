import { useState } from "react";
import './App.css';
import { GifGrid } from "./components";
import { AddCategory } from './components/AddCategory';
import { categories as cats } from "./data";

function App() {

  const [categories, setCategories] = useState(cats);

  return (
    <>
      <h2>Gif's App</h2>
      <hr />
      <AddCategory setCategories={setCategories} />
      {Boolean(categories.length) && categories.map((category, index) => (
        <GifGrid category={category} key={index + category} />
      ))
      }



    </>
  );
}

export default App;
