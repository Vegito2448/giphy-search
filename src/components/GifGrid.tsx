import 'animate.css';
import { useFetch } from "../hooks";
import { GifsResponse } from "../interfaces";
import { GifGridItem } from "./GifGridItem";

interface GifGridProps {
  category: string;
}

export const GifGrid = ({ category }: GifGridProps) => {
  const apiKey = import.meta.env.VITE_GIPHY_API_KEY;

  const url = `https://api.giphy.com/v1/gifs/search?q=${encodeURI(category)}&limit=10&api_key=${apiKey}`;

  const { data, loading } = useFetch<GifsResponse>({ url });

  const images = data?.data;

  return (
    <>
      <h3
        className="animate__animated animate__fadeIn"
      >{category}</h3>
      <div
        className="card-grid animate__animated animate__fadeIn animate__delay-2s"
      >
        {loading && <p
          className="animate__animated animate__flash animate__infinite animate__slower"
        >Loading...</p>}
        {!loading && images && Boolean(images.length) && <ol className="card-grid" >
          {images.map(({ id, title, images: { downsized_medium: { url } } }) => (
            <GifGridItem
              key={id}
              title={title}
              url={url}
              id={id}
            />
          ))}
        </ol>}
      </div >
    </>
  );
};
