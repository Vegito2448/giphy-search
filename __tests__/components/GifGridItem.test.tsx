import { render, screen } from '@testing-library/react';
import { GifGridItem } from "../../src/components";

describe('GifGridItem Component', () => {

  const props = {
    title: 'Title',
    url: 'https://localhost/image.jpg',
    id: '1'

  };

  test('matches snapshot', () => {
    const { asFragment } = render(<GifGridItem {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });

  beforeEach(() => {
    render(<GifGridItem
      {...props}
    />);
  });

  test('should render the component', () => {
    const titleElement = screen.getByText('Title');
    expect(titleElement).toBeInTheDocument();
  });

  test('should render the image with correct src and alt attributes', () => {
    const imageElement = screen.getByRole('img', { name: props.title });
    expect(imageElement).toBeInTheDocument();
    expect(imageElement).toHaveAttribute('src', props.url);
    expect(imageElement).toHaveAttribute('alt', props.title);
  });

  test('should render container with class animate__fadeIn', () => {
    const container = document.querySelector('.animate__fadeIn');
    expect(container).toBeInTheDocument();
  });

});