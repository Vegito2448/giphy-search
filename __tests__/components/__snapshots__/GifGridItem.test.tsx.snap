// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GifGridItem Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="card animate__animated animate__fadeIn animate__delay-2s"
  >
    <img
      alt="Title"
      src="https://localhost/image.jpg"
    />
    <p>
      Title
    </p>
  </div>
</DocumentFragment>
`;
